### ISSUE TITLE
Failed to start the server

### ISSUE THREAD

**Issue by oeway** [triton-inference-server/pytriton#1]
Hi, I am trying the example in README, but failed to start the server with the following code:
```
(cellpose) weiouyang@europa:~/workspace/pytriton$ python test-pytriton.py 
Triton Inference Server exited with failure. Please wait.
Internal proxy backend error. It will be closed.
Context was terminated
Traceback (most recent call last):
  File "/home/weiouyang/workspace/pytriton/pytriton/models/model.py", line 222, in _model_proxy_handshake
    socket.recv()
  File "zmq/backend/cython/socket.pyx", line 803, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 839, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 193, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/socket.pyx", line 188, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ContextTerminated: Context was terminated
Traceback (most recent call last):
  File "test-pytriton.py", line 34, in <module>
    triton.serve()
  File "/home/weiouyang/workspace/pytriton/pytriton/triton.py", line 343, in serve
    self.run()
  File "/home/weiouyang/workspace/pytriton/pytriton/triton.py", line 318, in run
    self._wait_for_models()
  File "/home/weiouyang/workspace/pytriton/pytriton/triton.py", line 429, in _wait_for_models
    client.wait_for_model(timeout_s=120)
  File "/home/weiouyang/workspace/pytriton/pytriton/client/client.py", line 154, in wait_for_model
    wait_for_model_ready(self._client, self._model_name, self._model_version, timeout_s=timeout_s)
  File "/home/weiouyang/workspace/pytriton/pytriton/client/utils.py", line 224, in wait_for_model_ready
    is_model_ready = condition.wait_for(_is_model_ready, timeout=min(1.0, timeout_s))
  File "/home/weiouyang/miniconda3/envs/cellpose/lib/python3.8/threading.py", line 328, in wait_for
    result = predicate()
  File "/home/weiouyang/workspace/pytriton/pytriton/client/utils.py", line 207, in _is_model_ready
    raise PytritonClientModelUnavailableError(f"Model {model_name}/{model_version_msg} is unavailable.")
pytriton.client.exceptions.PytritonClientModelUnavailableError: Model Linear/1 is unavailable.
```

I installed pytriton by build from source (on `main` branch, commit hash: `bc71f5840e909f58beaba5ecd3c4ea624e242523`).

Also a minor suggestion to improve the README is to add `from pytriton.decorators import batch` when using the `batch` decorator.

**Comment by jkosek** [comment]
Unfortunately I was not able to reproduce the issue. I was using Ubuntu 20.04.6 LTS.

Could you please download the latest wheel from https://github.com/triton-inference-server/pytriton/releases/tag/v0.1.4 and try again one of the examples?

We also recommend to use virtualenv:
```
python -m venv venv
source venv/bin/activate
```

**Comment by oeway** [comment]
Hi I just did another test with the following steps and failed:
```
conda create -n pytriton python=3.8
conda activate pytriton
pip install https://github.com/triton-inference-server/pytriton/releases/download/v0.1.4/pytriton-0.1.4-py3-none-manylinux_2_31_x86_64.whl
pip install torch
```

Here is the test file:
[test-pytriton.py.zip](https://github.com/triton-inference-server/pytriton/files/11020864/test-pytriton.py.zip)

The error message:
```
(pytriton) weiouyang@europa:~/workspace/pytriton$ python test-pytriton.py 
Triton Inference Server exited with failure. Please wait.
Internal proxy backend error. It will be closed.
Context was terminated
Traceback (most recent call last):
  File "/home/weiouyang/workspace/pytriton/pytriton/models/model.py", line 222, in _model_proxy_handshake
    socket.recv()
  File "zmq/backend/cython/socket.pyx", line 803, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 839, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 193, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/socket.pyx", line 188, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ContextTerminated: Context was terminated
Internal proxy backend error. It will be closed.
Context was terminated
Traceback (most recent call last):
  File "/home/weiouyang/workspace/pytriton/pytriton/models/model.py", line 222, in _model_proxy_handshake
    socket.recv()
  File "zmq/backend/cython/socket.pyx", line 803, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 839, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 193, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/socket.pyx", line 188, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ContextTerminated: Context was terminated
Traceback (most recent call last):
  File "test-pytriton.py", line 52, in <module>
    triton.serve()
  File "/home/weiouyang/workspace/pytriton/pytriton/triton.py", line 343, in serve
    self.run()
  File "/home/weiouyang/workspace/pytriton/pytriton/triton.py", line 318, in run
    self._wait_for_models()
  File "/home/weiouyang/workspace/pytriton/pytriton/triton.py", line 429, in _wait_for_models
    client.wait_for_model(timeout_s=120)
  File "/home/weiouyang/workspace/pytriton/pytriton/client/client.py", line 154, in wait_for_model
    wait_for_model_ready(self._client, self._model_name, self._model_version, timeout_s=timeout_s)
  File "/home/weiouyang/workspace/pytriton/pytriton/client/utils.py", line 224, in wait_for_model_ready
    is_model_ready = condition.wait_for(_is_model_ready, timeout=min(1.0, timeout_s))
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/threading.py", line 328, in wait_for
    result = predicate()
  File "/home/weiouyang/workspace/pytriton/pytriton/client/utils.py", line 207, in _is_model_ready
    raise PytritonClientModelUnavailableError(f"Model {model_name}/{model_version_msg} is unavailable.")
pytriton.client.exceptions.PytritonClientModelUnavailableError: Model LinearInfer/1 is unavailable.
```

**Comment by jkosek** [comment]
Ok, I see you are using conda. Let us check that. In the mean time, could you simply use `python` and `pip` commands?

**Comment by oeway** [comment]
Hi, sorry for the delay.

With the two commands, I found the issue is from the port 8000 is already occupied. Maybe we can improve the error log for pytriton? Instead of saying `Model LinearInfer/1 is unavailable.` maybe we can display the error log of triton server?


Here is the log:
```
(pytriton) weiouyang@europa:~/workspace/pytriton$ CUDA_VISIBLE_DEVICES=1 python /tmp/pytriton_server_test.py 
DEBUG:pytriton.utils.workspace:Workspace path /home/weiouyang/.cache/pytriton/workspace_udbldvb5
DEBUG:pytriton.models.manager:Adding LinearInfer (1) to registry under ('linearinfer', 1).
DEBUG:pytriton.models.manager:Adding LinearTrain (1) to registry under ('lineartrain', 1).
DEBUG:pytriton.models.manager:Crating model LinearInfer with version 1.
DEBUG:pytriton.models.model:Generating model and config for LinearInfer and 1 to /home/weiouyang/.cache/pytriton/workspace_udbldvb5/model-store
DEBUG:pytriton.model_config.generator:Generated Triton config:
{
    "name": "LinearInfer",
    "backend": "python",
    "max_batch_size": 128,
    "dynamic_batching": {},
    "input": [
        {
            "name": "INPUT_1",
            "dims": [
                -1
            ],
            "data_type": "TYPE_FP32"
        }
    ],
    "output": [
        {
            "name": "OUTPUT_1",
            "dims": [
                -1
            ],
            "data_type": "TYPE_FP32"
        }
    ],
    "instance_group": [
        {
            "count": 1,
            "kind": "KIND_CPU"
        }
    ],
    "parameters": {
        "shared-memory-socket": {
            "string_value": "ipc:///home/weiouyang/.cache/pytriton/workspace_udbldvb5/ipc_proxy_backend_LinearInfer"
        }
    }
}
DEBUG:pytriton.model_config.generator:Generated Triton config payload:
name: "LinearInfer"
max_batch_size: 128
input {
  name: "INPUT_1"
  data_type: TYPE_FP32
  dims: -1
}
output {
  name: "OUTPUT_1"
  data_type: TYPE_FP32
  dims: -1
}
instance_group {
  count: 1
  kind: KIND_CPU
}
dynamic_batching {
}
parameters {
  key: "shared-memory-socket"
  value {
    string_value: "ipc:///home/weiouyang/.cache/pytriton/workspace_udbldvb5/ipc_proxy_backend_LinearInfer"
  }
}
backend: "python"

DEBUG:pytriton.model_config.generator:Generated config stored in /home/weiouyang/.cache/pytriton/workspace_udbldvb5/model-store/LinearInfer/config.pbtxt
DEBUG:pytriton.proxy.inference_handler:Binding IPC socket at ipc:///home/weiouyang/.cache/pytriton/workspace_udbldvb5/ipc_proxy_backend_LinearInfer_0.
DEBUG:pytriton.proxy.inference_handler:Waiting for requests from proxy model for LinearInfer.
DEBUG:pytriton.models.manager:Done.
DEBUG:pytriton.models.manager:Crating model LinearTrain with version 1.
DEBUG:pytriton.models.model:Generating model and config for LinearTrain and 1 to /home/weiouyang/.cache/pytriton/workspace_udbldvb5/model-store
DEBUG:pytriton.model_config.generator:Generated Triton config:
{
    "name": "LinearTrain",
    "backend": "python",
    "max_batch_size": 128,
    "dynamic_batching": {},
    "input": [
        {
            "name": "INPUT_1",
            "dims": [
                -1
            ],
            "data_type": "TYPE_FP32"
        }
    ],
    "output": [
        {
            "name": "OUTPUT_1",
            "dims": [
                -1
            ],
            "data_type": "TYPE_FP32"
        }
    ],
    "instance_group": [
        {
            "count": 1,
            "kind": "KIND_CPU"
        }
    ],
    "parameters": {
        "shared-memory-socket": {
            "string_value": "ipc:///home/weiouyang/.cache/pytriton/workspace_udbldvb5/ipc_proxy_backend_LinearTrain"
        }
    }
}
DEBUG:pytriton.model_config.generator:Generated Triton config payload:
name: "LinearTrain"
max_batch_size: 128
input {
  name: "INPUT_1"
  data_type: TYPE_FP32
  dims: -1
}
output {
  name: "OUTPUT_1"
  data_type: TYPE_FP32
  dims: -1
}
instance_group {
  count: 1
  kind: KIND_CPU
}
dynamic_batching {
}
parameters {
  key: "shared-memory-socket"
  value {
    string_value: "ipc:///home/weiouyang/.cache/pytriton/workspace_udbldvb5/ipc_proxy_backend_LinearTrain"
  }
}
backend: "python"

DEBUG:pytriton.model_config.generator:Generated config stored in /home/weiouyang/.cache/pytriton/workspace_udbldvb5/model-store/LinearTrain/config.pbtxt
DEBUG:pytriton.proxy.inference_handler:Binding IPC socket at ipc:///home/weiouyang/.cache/pytriton/workspace_udbldvb5/ipc_proxy_backend_LinearTrain_0.
DEBUG:pytriton.proxy.inference_handler:Waiting for requests from proxy model for LinearTrain.
DEBUG:pytriton.models.manager:Done.
DEBUG:pytriton.client.client:Creating InferenceServerClient for http://127.0.0.1:8000
DEBUG:pytriton.client.utils:Waiting for server to be ready (timeout=119.9999725818634)
DEBUG:pytriton.client.utils:Waiting for model LinearInfer/1 to be ready (timeout=119.98813796043396)
DEBUG:pytriton.client.utils:Parsing model repository index entries:
DEBUG:pytriton.client.utils:    name=affable-shark version=1 state=READY
DEBUG:pytriton.client.utils:    name=bestfitting-densenet version=1 state=READY
DEBUG:pytriton.client.utils:    name=bestfitting-inceptionv3-single-cell version=1 state=READY
DEBUG:pytriton.client.utils:    name=bioengine-model-runner version=1 state=READY
DEBUG:pytriton.client.utils:    name=cellpose-predict version=1 state=READY
DEBUG:pytriton.client.utils:    name=cellpose-python version=1 state=READY
DEBUG:pytriton.client.utils:    name=cellpose-train version=1 state=READY
DEBUG:pytriton.client.utils:    name=chatty-frog version=1 state=READY
DEBUG:pytriton.client.utils:    name=conscientious-seashell version=1 state=READY
DEBUG:pytriton.client.utils:    name=creative-panda version=1 state=READY
DEBUG:pytriton.client.utils:    name=determined-chipmunk version=1 state=READY
DEBUG:pytriton.client.utils:    name=discreet-rooster version=1 state=READY
DEBUG:pytriton.client.utils:    name=fearless-crab version=1 state=READY
DEBUG:pytriton.client.utils:    name=hiding-blowfish version=1 state=READY
DEBUG:pytriton.client.utils:    name=hiding-tiger version=1 state=READY
DEBUG:pytriton.client.utils:    name=humorous-owl version=1 state=READY
DEBUG:pytriton.client.utils:    name=impartial-shrimp version=1 state=READY
DEBUG:pytriton.client.utils:    name=independent-shrimp version=1 state=READY
DEBUG:pytriton.client.utils:    name=kind-seashell version=1 state=READY
DEBUG:pytriton.client.utils:    name=loyal-parrot version=1 state=READY
DEBUG:pytriton.client.utils:    name=m2unet-predict version=1 state=READY
DEBUG:pytriton.client.utils:    name=m2unet-train version=1 state=READY
DEBUG:pytriton.client.utils:    name=naked-microbe version=1 state=READY
DEBUG:pytriton.client.utils:    name=non-judgemental-eagle version=1 state=READY
DEBUG:pytriton.client.utils:    name=organized-badger version=1 state=READY
DEBUG:pytriton.client.utils:    name=polite-pig version=1 state=READY
DEBUG:pytriton.client.utils:    name=powerful-chipmunk version=1 state=READY
DEBUG:pytriton.client.utils:    name=shivering-raccoon version=1 state=READY
DEBUG:pytriton.client.utils:    name=stardist version=1 state=READY
DEBUG:pytriton.client.utils:    name=straightforward-crocodile version=1 state=READY
DEBUG:pytriton.client.utils:    name=willing-hedgehog version=1 state=READY
DEBUG:pytriton.client.client:Closing InferenceServerClient
DEBUG:pytriton.triton:Stopping Triton Inference server and proxy backends
DEBUG:pytriton.server.triton_server:Stopping Triton Inference server - sending SIGINT signal and wait 30s
DEBUG:pytriton.server.triton_server:Waiting for process to stop
WARNING:pytriton.server.triton_server:Triton Inference Server exited with failure. Please wait.
DEBUG:pytriton.server.triton_server:Triton Inference Server exit code -2
DEBUG:pytriton.triton:Got callback that tritonserver process finished
DEBUG:pytriton.server.triton_server:Error message: 
Traceback (most recent call last):
  File "/tmp/pytriton_server_test.py", line 55, in <module>
    triton.serve()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/triton.py", line 347, in serve
    self.run()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/triton.py", line 322, in run
    self._wait_for_models()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/triton.py", line 436, in _wait_for_models
    client.wait_for_model(timeout_s=120)
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/client/client.py", line 160, in wait_for_model
    wait_for_model_ready(self._client, self._model_name, self._model_version, timeout_s=timeout_s)
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/client/utils.py", line 224, in wait_for_model_ready
    is_model_ready = condition.wait_for(_is_model_ready, timeout=min(1.0, timeout_s))
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/threading.py", line 328, in wait_for
    result = predicate()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/client/utils.py", line 207, in _is_model_ready
    raise PyTritonClientModelUnavailableError(f"Model {model_name}/{model_version_msg} is unavailable.")
pytriton.client.exceptions.PyTritonClientModelUnavailableError: Model LinearInfer/1 is unavailable.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/server/triton_server.py", line 178, in stop
    self._tritonserver_running_cmd.wait(timeout=SERVER_OUTPUT_TIMEOUT_SECS)
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/sh.py", line 845, in wait
    self.handle_command_exit_code(exit_code)
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/sh.py", line 869, in handle_command_exit_code
    raise exc
sh.SignalException_SIGINT: 

  RAN: /home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/tritonserver/bin/tritonserver --model-repository /home/weiouyang/.cache/pytriton/workspace_udbldvb5/model-store --backend-directory /home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/tritonserver/backends

  STDOUT:


  STDERR:


DEBUG:pytriton.models.manager:Clean model ('linearinfer', 1).
DEBUG:pytriton.models.model:Closing socket if needed
INFO:pytriton.proxy.inference_handler:Context was terminated. InferenceHandler will be closed.
ERROR:pytriton.models.model:Internal proxy backend error. It will be closed.
INFO:pytriton.proxy.inference_handler:Closing socket
ERROR:pytriton.models.model:Context was terminated
Traceback (most recent call last):
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/models/model.py", line 245, in _model_proxy_handshake
    socket.recv()
  File "zmq/backend/cython/socket.pyx", line 803, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 839, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 193, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/socket.pyx", line 188, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ContextTerminated: Context was terminated
INFO:pytriton.proxy.inference_handler:Closing buffers
INFO:pytriton.models.model:Closing socket
INFO:pytriton.proxy.inference_handler:Leaving proxy backend thread
INFO:pytriton.triton:Received ModelEvent.RUNTIME_TERMINATED from <pytriton.models.model.Model object at 0x7f82e8318bb0>; context=None
DEBUG:pytriton.models.model:Socket closed. Waiting for proxy backend to shut down
INFO:pytriton.proxy.inference_handler:Closing proxy
DEBUG:pytriton.models.model:All backends 
DEBUG:pytriton.models.manager:Clean model ('lineartrain', 1).
DEBUG:pytriton.models.model:Closing socket if needed
ERROR:pytriton.models.model:Internal proxy backend error. It will be closed.
INFO:pytriton.proxy.inference_handler:Context was terminated. InferenceHandler will be closed.
ERROR:pytriton.models.model:Context was terminated
Traceback (most recent call last):
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/models/model.py", line 245, in _model_proxy_handshake
    socket.recv()
  File "zmq/backend/cython/socket.pyx", line 803, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 839, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 193, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/socket.pyx", line 188, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ContextTerminated: Context was terminated
INFO:pytriton.proxy.inference_handler:Closing socket
INFO:pytriton.models.model:Closing socket
INFO:pytriton.proxy.inference_handler:Closing buffers
INFO:pytriton.proxy.inference_handler:Leaving proxy backend thread
INFO:pytriton.triton:Received ModelEvent.RUNTIME_TERMINATED from <pytriton.models.model.Model object at 0x7f82e8318df0>; context=None
DEBUG:pytriton.models.model:Socket closed. Waiting for proxy backend to shut down
INFO:pytriton.proxy.inference_handler:Closing proxy
DEBUG:pytriton.models.model:All backends 
DEBUG:pytriton.utils.workspace:Cleaning workspace dir /home/weiouyang/.cache/pytriton/workspace_udbldvb5
DEBUG:pytriton.triton:Stopped Triton Inference server and proxy backends
Traceback (most recent call last):
  File "/tmp/pytriton_server_test.py", line 55, in <module>
    triton.serve()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/triton.py", line 347, in serve
    self.run()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/triton.py", line 322, in run
    self._wait_for_models()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/triton.py", line 436, in _wait_for_models
    client.wait_for_model(timeout_s=120)
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/client/client.py", line 160, in wait_for_model
    wait_for_model_ready(self._client, self._model_name, self._model_version, timeout_s=timeout_s)
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/client/utils.py", line 224, in wait_for_model_ready
    is_model_ready = condition.wait_for(_is_model_ready, timeout=min(1.0, timeout_s))
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/threading.py", line 328, in wait_for
    result = predicate()
  File "/home/weiouyang/miniconda3/envs/pytriton/lib/python3.8/site-packages/pytriton/client/utils.py", line 207, in _is_model_ready
    raise PyTritonClientModelUnavailableError(f"Model {model_name}/{model_version_msg} is unavailable.")
pytriton.client.exceptions.PyTritonClientModelUnavailableError: Model LinearInfer/1 is unavailable.
```


Please feel free to close the issue.

**Comment by piotrm-nvidia** [comment]
I'd like to provide additional details regarding PyTriton version 0.5.0 in relation to the Triton server logs. Currently, when the server encounters a port conflict, it outputs an error message for the first port that fails to bind and ceases further error reporting. This means that if there are multiple conflicts with existing processes listening on those ports, only the first conflict is reported.

Here are the error messages for each type of service when they encounter a port already in use:


For the HTTP service:
```
E0118 08:31:52.149680 840 main.cc:257] failed to start HTTP service: Unavailable - Socket '0.0.0.0:1111' already in use
```

For the GRPC service:
```
E0118 08:25:14.316508 570 main.cc:245] failed to start GRPC service: Unavailable - Socket '0.0.0.0:2222' already in use
```

For the Metrics service:
```
E0118 08:33:40.673958 988 main.cc:293] failed to start Metrics service: Unavailable - Socket '0.0.0.0:3333' already in use
```

To reproduce these messages in your environment, you can use the following script. This server example demonstrates how to set up a Triton server instance with specific ports for HTTP, GRPC, and Metrics services.

Server example script:

```python
from pytriton.decorators import batch
from pytriton.model_config import ModelConfig, Tensor
from pytriton.triton import Triton, TritonConfig
import numpy as np

@batch
def infer_fn(**inputs: np.ndarray):
    return {"token": np.zeros((1,1), dtype=np.int64)}

MAX_BATCH_SIZE = 10

with Triton(
        config=TritonConfig(http_port=1111, grpc_port=2222, metrics_port=3333)
    ) as triton:
        triton.bind(
            model_name="Whisper",
            infer_func=infer_fn,
            inputs=[
                Tensor(name="audio_feature", dtype=np.float32, shape=(80, 3000)),
            ],
            outputs=[
                Tensor(name="token", dtype=np.int64, shape=(-1,)),
            ],
            config=ModelConfig(max_batch_size=MAX_BATCH_SIZE),
            strict=True,
        )
        triton.serve()
```

This code sets up a Triton server instance and binds a model named "Whisper" to specific input and output tensors, with a maximum batch size defined. The `serve` method is called to start serving the model.


Full log from failing Triton with conflict for all ports:

```
I0118 08:25:14.263362 237 model_lifecycle.cc:461] loading: Whisper:1
W0118 08:25:14.267811 570 pinned_memory_manager.cc:237] Unable to allocate pinned system memory, pinned memory pool will not be available: unknown error
E0118 08:25:14.267874 570 server.cc:236] Failed to initialize CUDA memory manager: unable to get number of CUDA devices: unknown error
E0118 08:25:14.267881 570 server.cc:243] unable to get number of CUDA devices: unknown error
W0118 08:25:14.267888 570 server.cc:251] failed to enable peer access for some device pairs
I0118 08:25:14.268128 570 server.cc:606] 
+------------------+------+
| Repository Agent | Path |
+------------------+------+
+------------------+------+

I0118 08:25:14.268143 570 server.cc:633] 
+---------+------+--------+
| Backend | Path | Config |
+---------+------+--------+
+---------+------+--------+

I0118 08:25:14.268155 570 server.cc:676] 
+-------+---------+--------+
| Model | Version | Status |
+-------+---------+--------+
+-------+---------+--------+

W0118 08:25:14.315066 570 metrics.cc:801] Cannot get CUDA device count, GPU metrics will not be available
I0118 08:25:14.315232 570 metrics.cc:710] Collecting CPU metrics
I0118 08:25:14.315356 570 tritonserver.cc:2483] 
+----------------------------------+------------------------------------------+
| Option                           | Value                                    |
+----------------------------------+------------------------------------------+
| server_id                        | triton                                   |
| server_version                   | 2.41.0                                   |
| server_extensions                | classification sequence model_repository |
|                                  |  model_repository(unload_dependents) sch |
|                                  | edule_policy model_configuration system_ |
|                                  | shared_memory cuda_shared_memory binary_ |
|                                  | tensor_data parameters statistics trace  |
|                                  | logging                                  |
| model_repository_path[0]         | /root/.cache/pytriton/workspace_tn248wzz |
| model_control_mode               | MODE_EXPLICIT                            |
| strict_model_config              | 0                                        |
| rate_limit                       | OFF                                      |
| pinned_memory_pool_byte_size     | 268435456                                |
| min_supported_compute_capability | 6.0                                      |
| strict_readiness                 | 1                                        |
| exit_timeout                     | 30                                       |
| cache_enabled                    | 0                                        |
+----------------------------------+------------------------------------------+

E0118 17:25:14.316253476     570 chttp2_server.cc:1080]      UNKNOWN:No address added out of total 1 resolved for '0.0.0.0:2222' {created_time:"2024-01-18T17:25:14.316165717+09:00", children:[UNKNOWN:Failed to add any wildcard listeners {created_time:"2024-01-18T17:25:14.316152831+09:00", children:[UNKNOWN:Unable to configure socket {fd:11, created_time:"2024-01-18T17:25:14.31611928+09:00", children:[UNKNOWN:Address already in use {syscall:"bind", os_error:"Address already in use", errno:98, created_time:"2024-01-18T17:25:14.316087637+09:00"}]}, UNKNOWN:Unable to configure socket {fd:11, created_time:"2024-01-18T17:25:14.316147962+09:00", children:[UNKNOWN:Address already in use {created_time:"2024-01-18T17:25:14.316144535+09:00", errno:98, os_error:"Address already in use", syscall:"bind"}]}]}]}
E0118 08:25:14.316508 570 main.cc:245] failed to start GRPC service: Unavailable - Socket '0.0.0.0:2222' already in use 
W0118 08:25:15.316113 570 metrics.cc:251] Unable to destroy DCGM group: Setting not configured
I0118 08:25:15.572568 237 python_be.cc:2363] TRITONBACKEND_ModelInstanceInitialize: Whisper_0_0 (CPU device 0)
I0118 08:25:15.959530 237 model_lifecycle.cc:818] successfully loaded 'Whisper'
Triton Inference Server exited with failure. Please wait.
E0118 08:25:17.002982 237 model_lifecycle.cc:408] Agent model returns error on TRITONREPOAGENT_ACTION_UNLOAD: Internal: Unexpected lifecycle state transition from TRITONREPOAGENT_ACTION_LOAD to TRITONREPOAGENT_ACTION_UNLOAD
I0118 08:25:19.274712 237 model_lifecycle.cc:603] successfully unloaded 'Whisper' version 1
```
